Lib/asyncio/proactor_events.py

```diff
@@ -280,11 +287,13 @@ def _loop_reading(self, fut=None):
                 if fut.done():
                     # deliver data later in "finally" clause
                     length = fut.result()
-                    if length == 0:
-                        # we got end-of-file so no need to reschedule a new read
-                        return
-
-                    data = self._data[:length]
+                    if length > -1:
+                        if not isinstance(self._protocol, protocols.BufferedProtocol):
```

Comment:
Do we have proper functional tests that hit this branch? If not we need to add them.

Reply:
You can probably copy some tests from the uvloop project (as I own the project there will be no licensing issues even if it's a literal copy)

Reply:
I am thinking of doing parameterize on all the appropriate test in test_proactor_events.py.

That way we can use the already existing tests on both Protocol and BufferedProtocol

Reply:
Ok I did this using the parameterize function, it did made me a bit of ugly code I want you to look at.

Check out what I did in the test adding commit.

By the way shouldn't I add the same tests for ``UnixReadPipeTransportTests``?

Reply:
Ok I added the tests to UnixReadPipeTransportTests as well using the same patches

Reply:
> Ok I did this using the parameterize function, it did made me write a bit of ugly code I want you to look at.

Yeah, let's try to simplify it. See my other comment.

> By the way shouldn't I add the same tests for UnixReadPipeTransportTests?

I guess since we're updating `_UnixReadPipeTransport` we should.

Reply:
Yeah I simplified it a lot